### PR TITLE
New Meta Design

### PR DESCRIPTION
**Pull request by ehsteve**
Does not include any code but shows the API for a new way to handle meta data. This would underly the changes that were made by @Alex-Ian-Hamilton in the TimeSeries branch.

### CODE REVIEW

**Comment by Cadair** [comment]
Can you expand on the justification for this? What problem is it trying to solve?

I have been thinking about adding a more detailed Meta object at the `sunpy.io` level this week to solve some of the issues @Alex-Ian-Hamilton has been having with GOES data and the database, this doesn't address those issues, but I think it might be a good starting point.

**Comment by ehsteve** [comment]
The point of this is to provide a consistent and foundational interface to metadata across all sunpy objects that include the ability to see how the metadata has changed and verify that it contains minimum information needed.

**Comment by Cadair** [comment]
I agree having a history tracker is important and I think this is a great place for it, I am cautious about the idea of "minimum information" as to be that feels like more a job for something using the meta to decide.

**Comment by Cadair** [comment]
On a related note, I have been thinking about the idea of instrument specific metadata collections. The objective here being if they had something like this class as a base and then could provide instrument specific logic on top.

The main motivation for this right now is what @Alex-Ian-Hamilton has been doing with GOES files in the database, over the lifetime of "GOES" there has been many different ways of encoding time in the header, these are only decodeable with the contextual information of GOES (otherwise they are ambiguous) so to me they fit in something like a GOES specific meta data object.

I was thinking of having a collection class like this that has extra properties (like `date_obs`) that could be defined and accessed *if* they exist by classes like the database (otherwise falling back to dict style lookup). Thus allowing instrument specific metadata decoding to be implemented at the lowest common level of sunpy (sunpy.io) and therefore be utilised by all the parts of SunPy.

This train of thought got me onto the idea of having metadata abstract base classes, classes that form specifications of what properties should be defined for them to be valid for certain use cases. The most obvious of these being a metadata class that fulfils the FITS-WCS specification (i.e. has all of CTYPE, CDELT etc.), these things could also be defined as properties with (instrument specific) overrides and then a using class (i.e. map) would only have to assert that the metadata object it was getting was a subclass of `FITSWCSMetaData` to know it had implemented the correct interface.

File: sunpy/meta/meta.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
+from __future__ import absolute_import
+
+import numpy as np
+from astropy.units import Quantity
+
+__all__ = ['MetaItem', 'Meta']
+
+
+class Meta(object):
```
Comment (Cadair) [inline sunpy/meta/meta.py]: This object should subclass from the `collections.MutableMapping` abstract base class: https://docs.python.org/3.6/library/collections.abc.html

File: sunpy/meta/meta.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        """
+        Two Headers are equal only if they have the exact same string
+        representation.
+        """
+        return str(self) == str(other)
+
+    def __add__(self, meta):
+        """Equivalent to concatenate"""
+
+    @property
```
Comment (Cadair) [inline sunpy/meta/meta.py]: to match dictionary this probably should be a method and not a property?

File: sunpy/meta/meta.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    def items(self):
+        """
+        The underlying data that make up this Meta; it can be
+        looked at, but it should not be modified directly.
+        """
+        return self._items
+
+    @property
```
Comment (Cadair) [inline sunpy/meta/meta.py]: same as items

File: sunpy/meta/meta.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    def keywords(self):
+        """Return all of the keywords for all of the items"""
+
+    @property
```
Comment (Cadair) [inline sunpy/meta/meta.py]: same as items

File: sunpy/meta/meta.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    def fromDict(cls, data):
+        """Create a MetaData from a dictionary."""
+
+    def to_string(cls):
+        """Returns a string representation of the header"""
+
+    def to_FITS(self):
+        """Returns a FITS header"""
+
+    @classmethod
```
Comment (Cadair) [inline sunpy/meta/meta.py]: classmethod?

File: sunpy/meta/meta.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+        """
+        del self[self.index(keyword)]
+
+    def is_compliant(self):
+        """Check if the header is compliant (i.e. does it contain the minimum set of
+        items with keywords and by checking that each item is verified."""
+
+
+class MetaItem(object):
```
Comment (Cadair) [inline sunpy/meta/meta.py]: having this class means a fundamental shift away from the
```python
>>> mymeta['spam']
'eggs'
```
approach of getting values out of a indexing operation (much like the confusion you felt about TSMD)

Comment (Cadair) [inline sunpy/meta/meta.py]: Yeah true.

File: sunpy/meta/meta.py (hunk 1 above)
Comment (ehsteve) [inline sunpy/meta/meta.py]: Great idea!

**Comment by dpshelio** [comment]
Is this something like having a metadata model that it works for anything? [spacepy has something like that](https://pythonhosted.org/SpacePy/datamodel.html)... Is this following the same line of thought?

**Comment by Raahul-Singh** [comment]
What more needs to be done here?

**Pull request closed without merging**